### DISCUSSION TITLE
Rule Expressions and short circuit evaluation

### DISCUSSION THREAD

**Discussion by bmasotta-lcg** [microsoft/RulesEngine#446] [Q&A] [answered]
I found that for the following expression, the engine tries to evaluate the entire expression instead of performing a short circuit evaluation. Is this by design?

Expression: 
"input1.falseProp == true AND input1.missingProp"

Error Returned:
"Exception while parsing expression `input1.falseProp == true AND input1.missingProp` - Operator 'AND' incompatible with operand types 'Boolean' and 'Object'"

InputMessage:
{
    "testInputName": "shortCircuitMissingProperty",
    "falseProp": false
}

notice there's no "missingProp" property in input1 message
My idea is to handle different input message types (schemas) with several rules, all in the same workflow, by filtering by a message property: 

input1.telemetryType == 'weather' AND input1.windSpeed > 100
...
input1.telemetryType == 'tracker' AND input1.motorCurrent > 5

Thanks a lot!

**Comment by bmasotta-lcg** [comment]
also I was looking to be able to handle short circuit evaluation by using nested rules and the OrElse operator... 
but is not working for me
having two child rules with a different custom action each one, both custom actions gets executed, no matter if the first one is true, the second one also gets evaluated...

**Comment by abbasc52** [comment, marked as answer]
@bmasotta-lcg Short circuiting using nested rules works only if you enable perf mode in ReSettings. You can refer this example.
https://github.com/microsoft/RulesEngine/blob/7b4e45d2addeba1abe8c6f8c4a38d6816077fab4/test/RulesEngine.UnitTest/NestedRulesTest.cs#L21-L65

Also, on your earlier question, if you could provide me more details on how you are passing the input (and the type of it). I might be able to help you better.

**Reply by bmasotta** [reply]
thanks @abbasc52 . I was able to implement sce by using nested rules. Also found it's working with static types but not dynamic type.
I need to investigate a little more. But it's weird, when using dynamic type all the IEnumerable logic is lost (I think it's related to the linq expression library implementation)

e.g: if I use static type collection methods in a CustomType (e.g.: IEnumerable<Telemetry> GetTelemetry()) there's no problem in using expressions as .Count(). Select(x=> and so on)
but if I switch to dynamic types (e.g.: IEnumerable<dynamic> GetTelemetry() ) I start to experience problems.

the thing is that for IoT solutions is very appreciated to be able to handle dynamic types (changin schemas) in a natural way

**Comment by bmasotta-lcg** [comment]
I was unable to handle missing or null values on dynamic types without using nested rules or separate workflows.
I created a sample repository showing this.
https://github.com/bmasotta-lcg/RulesEngineNullShortCircuit

While I create rules that look for always-present properties on dynamic type inputs I don't have any problem (that's similar to work with static types)

But if I need to work with dynamic or changing schemas in dynamic input messages and test for null on a specific field value, by using any kind of short circuit evaluation method, the RulesEngine will raise an error related to the expression evaluation. 

      "Expression": "input1.value != null && input1.value == 10",
 },
    "ExceptionMessage": "Exception while parsing expression \"input1.value != null && input1.value == 10\" - The binary operator Equal is not defined for the types \u0027System.Object\u0027 and \u0027System.Int32\u0027."

 "Expression": "input1.value != null AndAlso input1.value == 10",
"ExceptionMessage": "Exception while parsing expression \u0060input1.value != null AndAlso input1.value == 10\u0060 - The binary operator Equal is not defined for the types \u0027System.Object\u0027 and \u0027System.Int32\u0027."

(https://dynamic-linq.net/advanced-null-propagation)
 "Expression": "np(input1.value, 0) == 10",
 "ExceptionMessage": "Exception while parsing expression \u0060input1.value != null \u0026\u0026 input1.value == 10\u0060 - The binary operator Equal is not defined for the types \u0027System.Object\u0027 and \u0027System.Int32\u0027."


 "Expression": "(input1.value == null ? 0 : input1.value) == 10",
"ExceptionMessage": "Exception while parsing expression \u0060(input1.value == null ? 0 : input1.value) == 10\u0060 - The binary operator Equal is not defined for the types \u0027System.Object\u0027 and \u0027System.Int32\u0027."

One thing to add is that I find very confusing mentioning in the readme file that the rule expressions are similar to regular c# lambda expression. This is not true. The RulesEngine is silently using the https://dynamic-linq.net library, a library that looks like official .NET due to the namespace it uses, but it has it's own syntax, not the same syntax as for lamba expressions

**Reply by SolzIT** [reply]
Any updates on this error. In my case I always have the attribute but it is either null or has a integer value. For whatever methods I try I am getting the same error   

The binary operator Equal is not defined for the types System.Object and System.Int32."
